### ISSUE TITLE
Allow setting counter value when exposing existing counts

### ISSUE THREAD

**Issue by gambrose** [andrasm/prometheus-net#32]
I am just started to use prometheus, and your library and have come across this issue.

Counter only allows you to increase the value which is fine if you actually using it to count things but breaks down if you have an existing counter you are trying to expose though prometheus.

E.g. the DotNetStatsCollector seems to increment based on a calculated diff.
`_cpuTotal.Inc(_process.TotalProcessorTime.TotalSeconds - _cpuTotal.Value);`
This looks like a race condition to me as I cannot see anything stopping two scrapes occurring at the same time. In fact as I understand it multiple parallel scrapes is the preferred way to achive high availability of the prometheus server.

**Comment by andrasm** [comment]
Hi gambrose,

While technically that particular line is indeed subject to races, in reality it shouldn't matter as the prometheus server won't scrape the _same_ process in parallel, unless it's misconfigured (if that's possible at all).

What exactly are you trying to do?

**Comment by gambrose** [comment]
I was thinking of two servers running in parallel. E.g. main server and a second server for redundancy.

My scenario is to expose though prometheus an existing counter maintained by a framework I am using. I want to expose a counter and not a gauge as it is a counter.
In my scenario the update would no be happening during a scrap but on a thread managed by the framework. While I am fairly confident that there will not be a race condition in my situation having to do calculations to fit the api did not feel right.

**Comment by gambrose** [comment]
From the go client https://github.com/prometheus/client_golang/blob/master/prometheus/counter.go
```
// Deprecated: Use NewConstMetric to create a counter for an external
// value. A Counter should never be set.
Set(float64)
```

This seems to suggest that they use a different type of metric to expose existing counters. Do you have a similar mechanism to support this use case in your lib?

**Comment by godefroi** [comment]
#63 may be a duplicate of this?

**Comment by gambrose** [comment]
@godefroi It might be worth explaining a bit why your counters are being reset.
Counters should be constantly increasing and will only be reset on a restart of the process/machine. If the value can go back to zero without a restart then it is more likely should be mapped to a Gauge.

This issue is about exposing existing counters though to Prometheus though prometheus-net.

If you are say publishing counters that are created and reset in a separate process then I would say it is a duplicate.

**Comment by godefroi** [comment]
I'm exposing a Windows performance counter, which definitely behaves like a counter; it increases, unless the process is restarted, at which time, it resets to zero. I have no control over when this happens, and when it does, I cannot reset the Prometheus `Counter` instance to zero.

**Comment by sandersaares** [comment]
It is indeed the case that CpuTotal metric is best expressed as a gauge and not a counter, not only to avoid the potential race condition (the practical effects of which are likely insignificant and rare) but also to match the intent of the various metric types - the very fact that something only increases does not make it a counter. I will perform this change in version 2.0 of the library for the built-in .NET performance metrics.

In general, though, enabling counters to reset to zero does not appear to serve a purpose. Counters are meant for counting events in the code being instrumented. Events do not un-happen.

The Prometheus metrics APIs are really meant to be viewed from the point of view of the thing being instrumented. Quoting [Prometheus docs](https://prometheus.io/docs/instrumenting/clientlibs/) with my emphasis:

> Before you can monitor your services, you need to add instrumentation to **their code** via one of the Prometheus client libraries.

If you are coyping metrics from some external non-Prometheus instrumentation (e.g. Windows performance counters or external services) you are no doubt going to run into concept/terminology mismatches.

The Counter API is designed for counting events occurring in your own code and this does not require any reset/set functionality. If you wish to preserve the "Counter" type in the wire format while still being able to arbitrarily set a value from an external source, you are best served by implementing a custom Collector.

Closing this as by design (for the general case). The specific `DotNetStatsCollector` metric referenced in the original post will be changed to a Gauge in v2.0.

**Comment by gambrose** [comment]
> the very fact that something only increases does not make it a counter

I would argue that this is precisely what makes it a counter.

From the Prometheus docs:

> A counter is a cumulative metric that represents a single numerical value that only ever goes up.

Exposing CpuTotal as a gauge on the wire is wrong as it behaves as a counter there for it should be labeled as a counter. Forcing it into a gauge as this is what the lib wants is bad from the point of view of the metrics consumer.

I think having a separate api that is for exposing existing counters makes sense to avoid confusion in the main .net counter api. It would be nice if this was build into the lib and the lib used this to expose the counters it already exposes by default.

**Comment by godefroi** [comment]
I agree with @gambrose that representing process_cpu_seconds_total as a gauge is lying about what it is and how it behaves. By labeling it as a gauge, you're leaking your implementation detail to Prometheus and all the consumers that might want to use the value.

Instead of changing the `Counter` class to add a method to reset it to zero, adding a new class that makes it easier to expose imperfect sources correctly is perfectly adequate.

**Comment by sandersaares** [comment]
Are you familiar with what other Prometheus exporter implementations do to enable such scenarios? I would not want to invent something entirely prometheus-net specific here but if you can point to some equivalent behavior in an official Prometheus client library that would be a strong argument in favor of such features.

**Comment by godefroi** [comment]
I have no experience with any other Prometheus client libraries. I think that #73 expresses a sufficient resolution; if it were clear how to create a custom class that could report to Prometheus that it was a counter, but in fact behave from an API perspective as a gauge, then I would be satisfied. Windows performance counters are the obvious use case here, but it's quite possible that there are others. While I would love to rewrite all my software to expose metrics directly instead of relying on performance counters, in reality I have >150 environments all running various versions of my products, and updating each version to expose metrics to Prometheus is a gargantuan task that will realistically take years. I would like to use Prometheus as best I can in the mean time.

**Comment by gambrose** [comment]
@sandersaares I took a look at the go client (I do not use go so I could be wrong).
They had a set method on counter but it was removed in favour of using a ConstantMetric that is just used when creating a custom collector.

https://github.com/prometheus/client_golang/blob/b49b54cdb5f9850eb4c0090e5f9a8b3fc5113d5d/prometheus/value.go#L81

**Comment by sandersaares** [comment]
I have added an example on how to expose custom metrics as part of #73. Hopefully this enables some of the desired flexibility in this area.
